### PR TITLE
Consolidate API object code

### PR DESCRIPTION
**Pull request by TurkeyMcMac**
This PR reduces repetition in the code that implements userdata objects such as `VoxelManip` and `Settings`. Overall, it removes over 600 lines of code. It should also decrease the size of the executable, but I haven't measured that.

## To do

This PR is Ready for Review.

## How to test

There are many changes in the PR, and it's possible that some are erroneous. Someone should look through them to make sure this is not the case.

### CODE REVIEW

**Comment by sfan5** [comment]
Hmm, not sure if I like the closure/upvalue solution even if it is a bit faster. But we'll get to this when the other PR is done.

**Comment by TurkeyMcMac** [comment]
I probably won't rebase this unless the concept is approved.

**Comment by TurkeyMcMac** [comment]
Nevermind, this rebase at least was easy.

**Comment by sfan5** [comment]
Okay so like I hinted at I'm not a fan of the extra complexity of the upvalue/closure solution and I think we should keep using `luaL_checkudata`.
In that case what should be done to reduce duplication is to turn `checkobject` into a templated helper and the same `registerClass` change as already done now.

**Comment by TurkeyMcMac** [comment]
I have changed `checkObject` to be templated. That does clean up the code a bit. I have not removed `methodCheckObject` at the moment.

I should do a benchmark using PUC Lua where a RNG is used in a loop. This is a realistic situation to benchmark method overhead.

**Comment by TurkeyMcMac** [comment]
Realistic benchmark using PUC Lua:

```lua
local rng = PcgRandom(123)
local results = {}
local before = core.get_us_time()
for i = 1, 10000 do
	results[i] = rng:next()
end
local after = core.get_us_time()
core.debug("Benchmark:", after - before)
```

Time with `methodCheckObject`: ~1200
Time with `checkObject`: ~2100

**Comment by Desour** [comment]
All that `luaL_checkudata` does different is where it gets the class's userdata from:
https://github.com/minetest/minetest/blob/b1233056b76aa803123cc304d323887ad5fdfbae/lib/lua/src/lauxlib.c#L124-L137



You could try to store the userdatas in an array table in the registry instead of using a string index.

**Comment by TurkeyMcMac** [comment]
Good idea.

**Comment by sfan5** [comment]
This really should not be micro-optimized, with the numbers from the earlier benchmark the idiomatic solution with `luaL_checkudata` is 40 <i>nano</i>seconds slower than upvalues.

**Comment by TurkeyMcMac** [comment]
Probably true. In any case, optimization is no longer a goal of this PR.

File: src/script/lua_api/l_metadata.cpp [hunk 1]
```diff
@@ -286,13 +285,29 @@ bool MetaDataRef::handleFromTable(lua_State *L, int table, IMetadata *meta)
...
+		const luaL_Reg *methods)
+{
+	const luaL_Reg metamethods[] = {
+		{"__eq", l_equals},
+		{"__gc", gc_object},
+		{0, 0}
+	};
+	registerClass(L, name, methods, metamethods);
+
+	luaL_getmetatable(L, name);
```
Comment (sfan5) [inline src/script/lua_api/l_metadata.cpp:309]: should have a comment briefly describing the purpose

**Review by sfan5 — APPROVED** [review]
LGTM
if you've tested it reasonably feel free to merge

**Pull request merged**